File: Session_07_GraphBasics/multi_language_implementations/graph_representations.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

const V = 4 // Number of vertices

// 1. Adjacency Matrix Representation
func createAdjacencyMatrix(adjMatrix *[V][V]int) {
	// Initialize matrix with 0s
	for i := 0; i < V; i++ {
		for j := 0; j < V; j++ {
			adjMatrix[i][j] = 0
		}
	}

	// Add edges
	adjMatrix[0][1] = 1
	adjMatrix[0][2] = 1
	adjMatrix[1][3] = 1
	adjMatrix[3][0] = 1
}

func printAdjacencyMatrix(adjMatrix *[V][V]int) {
	fmt.Println("Adjacency Matrix:")
	for i := 0; i < V; i++ {
		for j := 0; j < V; j++ {
			fmt.Printf("%d ", adjMatrix[i][j])
		}
		fmt.Println()
	}
	fmt.Println()
}

// 2. Adjacency List Representation
type AdjListNode struct {
	dest int
	next *AdjListNode
}

type AdjList struct {
	head *AdjListNode
}

type Graph struct {
	numVertices int
	array       []AdjList
}

// Create a graph of V vertices
func createGraph(numVertices int) *Graph {
	g := &Graph{
		numVertices: numVertices,
		array:       make([]AdjList, numVertices),
	}
	for i := 0; i < numVertices; i++ {
		g.array[i].head = nil
	}
	return g
}

// Add edge to graph (Directed)
func addEdge(graph *Graph, src, dest int) {
	newNode := &AdjListNode{dest: dest}
	// Insert at head of the list for src
	newNode.next = graph.array[src].head
	graph.array[src].head = newNode
}

func printGraphList(graph *Graph) {
	fmt.Println("Adjacency List:")
	for v := 0; v < graph.numVertices; v++ {
		curr := graph.array[v].head
		fmt.Printf("  Vertex %d: ", v)
		for curr != nil {
			fmt.Printf("-> %d ", curr.dest)
			curr = curr.next
		}
		fmt.Println("-> NULL")
	}
	fmt.Println()
}

func main() {
	fmt.Println("=== Graph Representations in Go ===")
	fmt.Println()

	// Adjacency Matrix
	var adjMatrix [V][V]int
	createAdjacencyMatrix(&adjMatrix)
	printAdjacencyMatrix(&adjMatrix)

	// Adjacency List
	graph := createGraph(V)
	addEdge(graph, 0, 1)
	addEdge(graph, 0, 2)
	addEdge(graph, 1, 3)
	addEdge(graph, 3, 0)
	printGraphList(graph)

	fmt.Print("Press Enter to exit...")
	bufio.NewReader(os.Stdin).ReadBytes('\n')
}

File: Session_07_GraphBasics/multi_language_implementations/graph_representations.py
# Session 07: Graph Representations in Python
# Shows how to represent a graph using:
# 1. Adjacency Matrix
# 2. Adjacency List

V = 4  # Number of vertices

# 1. Adjacency Matrix Representation
def create_adjacency_matrix():
    adj_matrix = [[0] * V for _ in range(V)]
    
    # Add edges
    adj_matrix[0][1] = 1
    adj_matrix[0][2] = 1
    adj_matrix[1][3] = 1
    adj_matrix[3][0] = 1
    return adj_matrix

def print_adjacency_matrix(adj_matrix):
    print("Adjacency Matrix:")
    for i in range(V):
        for j in range(V):
            print(f"{adj_matrix[i][j]} ", end="")
        print()
    print()

# 2. Adjacency List Representation
class AdjListNode:
    def __init__(self, dest):
        self.dest = dest
        self.next = None

class Graph:
    def __init__(self, num_vertices):
        self.num_vertices = num_vertices
        self.head_array = [None] * num_vertices

    # Add edge to graph (Directed)
    def add_edge(self, src, dest):
        new_node = AdjListNode(dest)
        
        # Insert at head of the list for src
        new_node.next = self.head_array[src]
        self.head_array[src] = new_node

def print_graph_list(graph):
    print("Adjacency List:")
    for v in range(graph.num_vertices):
        curr = graph.head_array[v]
        print(f"  Vertex {v}: ", end="")
        while curr:
            print(f"-> {curr.dest} ", end="")
            curr = curr.next
        print("-> NULL")
    print()

def main():
    # Adjacency Matrix
    adj_matrix = create_adjacency_matrix()
    print_adjacency_matrix(adj_matrix)

    # Adjacency List
    graph = Graph(V)
    graph.add_edge(0, 1)
    graph.add_edge(0, 2)
    graph.add_edge(1, 3)
    graph.add_edge(3, 0)
    print_graph_list(graph)

if __name__ == "__main__":
    main()
